contracts/command-batcher/src/encoding/evm.rs

```diff
@@ -0,0 +1,483 @@
+};
+
+use super::traits;
+
+impl TryFrom<Message> for Command {
+    type Error = ContractError;
+
+    fn try_from(msg: Message) -> Result<Self, Self::Error> {
+        Ok(Command {
+            command_type: CommandType::ApproveContractCall, // TODO: this would change when other command types are supported
+            command_params: command_params(
+                msg.source_chain,
+                msg.source_address,
+                ethereum_types::Address::from_str(&msg.destination_address).map_err(|_| {
+                    ContractError::InvalidMessage {
+                        reason: "destination_address is not a valid EVM address".into(),
+                    }
+                })?,
+                msg.payload_hash.as_slice().try_into().map_err(|_| {
+                    ContractError::InvalidMessage {
+                        reason: "payload_hash is not a valid keccak256 hash".into(),
+                    }
+                })?,
+            ),
+            id: command_id(msg.id),
+        })
+    }
+}
+
+pub struct Builder;
+impl traits::Builder for Builder {
+    fn build_batch(
+        messages: Vec<Message>,
+        destination_chain_id: Uint256,
+    ) -> Result<CommandBatch, ContractError> {
+        let message_ids: Vec<String> = messages.iter().map(|msg| msg.id.clone()).collect();
+
+        let data = Data {
+            destination_chain_id,
+            commands: messages
+                .into_iter()
+                .map(|msg| msg.try_into())
+                .collect::<Result<Vec<Command>, ContractError>>()?,
+        };
+        let encoded_data = Encoder::encode_data(&data);
+
+        let id = batch_id(&message_ids);
+        let msg_to_sign = msg_to_sign(&encoded_data);
+
+        Ok(CommandBatch {
+            id,
+            message_ids,
+            data,
+            msg_to_sign,
+        })
+    }
+
+    fn build_proof(
+        snapshot: Snapshot,
+        signers: HashMap<String, Signature>,
+        pub_keys: HashMap<String, multisig::types::PublicKey>,
+    ) -> Result<Proof, ContractError> {
+        let mut operators = snapshot
+            .participants
+            .iter()
+            .map(|(_, participant)| {
+                let pub_key = pub_keys
+                    .get(&participant.address.to_string())
+                    .ok_or(ContractError::PublicKeyNotFound {
+                        participant: participant.address.to_string(),
+                    })?
+                    .into();
+
+                Ok(Operator {
+                    address: evm_address(pub_key)?,
+                    weight: participant.weight.into(),
+                    signature: signers.get(participant.address.as_str()).cloned(),
+                })
+            })
+            .collect::<Result<Vec<Operator>, ContractError>>()?;
+        operators.sort_by(|a, b| a.address.cmp(&b.address));
+
+        Ok(Proof {
+            operators,
+            threshold: snapshot.quorum.into(),
+        })
+    }
```

Comment:
Can we somehow ensure that every `Proof` always has the operators and signatures sorted like this? Maybe make a constructor?

Reply:
removed traits completely and added `new` constructors for types